File: C++/Count the nodes at distance K from leaf.cpp
class Solution
{
    public:
    vector<Node*> leaf;
    unordered_map<Node*,int> vis;
    void parenting(Node* root,unordered_map<Node*,Node*> &parent){
        if(root->left){
            parent[root->left] = root;
            parenting(root->left,parent);
        }
        if(root->right){
            parent[root->right] = root;
            parenting(root->right,parent);
        }
        if(root->left == NULL && root->right == NULL){
            leaf.push_back(root);
        }
    }
    int check(Node* leaf,unordered_map<Node*,Node*> &parent,int k){
        if(leaf == NULL){
            return 0;
        }
        if(k == 0 && vis[leaf] ==0 ){
            vis[leaf]=1;
            return 1;
        }
        if(k<0){
            return 0;
        }
        k--;
        check(parent[leaf],parent,k);
    }
    int printKDistantfromLeaf(Node* root, int k)
    {
        unordered_map<Node*,Node*> parent;
        parent[root] = NULL;
        parenting(root,parent);
        int result = 0;
        for(auto j: leaf){
            result += check(j,parent,k);
        }
        return result;
    }
};

File: JAVA/Count the nodes at distance K from leaf.java
class Solution
{
    int printKDistantfromLeaf(Node root, int k)
    {
        if(root == null){
            return 0;
        }
        Queue<Node> q = new ArrayDeque<>();
        q.add(root);
        int c = 0;
        while(!q.isEmpty()){
            int size = q.size();
            while(size-- > 0){
                Node curr = q.remove();
                if(solve(curr,k)){
                    c++;
                }
                if(curr.left != null){
                    q.add(curr.left);
                }
                if(curr.right != null){
                    q.add(curr.right);
                }
            }
        }
        return c;
    }
    public static boolean solve(Node root,int k){
        if(root == null){
            return false;
        }
        if((root.left  == null && root.right == null) && k == 0){
            return true;
        }
        if((root.left  == null && root.right == null) || k <= 0){
            return false;
        }
        boolean l = solve(root.left,k-1);
        boolean r = solve(root.right,k-1);
        return l || r;
        
    }

}
